compiler/optimizer/VPHandlers.cpp

```diff
@@ -1160,22 +1159,52 @@ TR::Node *constrainAnyIntLoad(OMR::ValuePropagation *vp, TR::Node *node)
       {
       TR::Node *array = node->getFirstChild()->getFirstChild();
       TR::Node *index = node->getFirstChild()->getSecondChild();
-      if (index->getOpCode().isLoadConst() && array->getOpCode().isIndirect())
+      if (index->getOpCode().isLoadConst())
          {
          bool isGlobal;
-         TR::Node* base = array->getFirstChild();
-         TR::VPConstraint *baseVPConstraint = vp->getConstraint(base, isGlobal);
-         if (baseVPConstraint && baseVPConstraint->isConstString())
+         if (array->getOpCode().isIndirect())
             {
-            TR::VPConstString *constString = baseVPConstraint->getClassType()->asConstString();
+            TR::Node* base = array->getFirstChild();
+            TR::VPConstraint *baseVPConstraint = vp->getConstraint(base, isGlobal);
+            if (baseVPConstraint)
+               {
+               if (baseVPConstraint->isConstString())
+                  {
+                  TR::VPConstString *constString = baseVPConstraint->getClassType()->asConstString();
 
-            uint16_t ch = constString->charAt(((TR::Compiler->target.is64Bit() ? index->getLongIntLow() : index->getInt())
-                                               - TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) / 2, vp->comp());
-            if (ch != 0)
+                  uint16_t ch = constString->charAt(((TR::Compiler->target.is64Bit() ? index->getLongIntLow() : index->getInt())
+                                                     - TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) / 2, vp->comp());
+                  if (ch != 0)
+                     {
+                     vp->replaceByConstant(node, TR::VPIntConst::create(vp, ch), isGlobal);
+                     return node;
+                     }
+                  }
+               }
+            }
+         TR::VPConstraint *arrayVPConstraint = vp->getConstraint(array, isGlobal);
+         if (arrayVPConstraint)
+            {
+#ifdef J9_PROJECT_SPECIFIC
+            TR::KnownObjectTable *knot = vp->comp()->getKnownObjectTable();
+            TR::VPKnownObject *kobj = arrayVPConstraint->getKnownObject();
+            if (knot && kobj)
                {
-               vp->replaceByConstant(node, TR::VPIntConst::create(vp, ch), true);
-               return node;
+               TR::KnownObjectTable::Index idx = kobj->getIndex();
+               if (kobj->isArrayWithConstantElements(vp->comp()))
+                  {
+                  TR::VMAccessCriticalSection constrainAnyIntLoadCriticalSection(vp->comp(),
+                     TR::VMAccessCriticalSection::tryToAcquireVMAccess);
+                  if (constrainAnyIntLoadCriticalSection.hasVMAccess())
+                     {
+                     uintptrj_t array = knot->getPointer(idx);
+                     uint16_t ch = *((uint16_t*)(array + (TR::Compiler->target.is64Bit() ? index->getLongIntLow() : index->getInt())));
```

Comment:
It's necessary to check bounds here. The load could be out of bounds if it's in a dead path. Also, the offset (`index`) shouldn't be truncated. I suggest doing all of the offset/index calculations in `uintptrj_t`. Don't take the `isConstString()` case above as a good example; it's unclear, and it looks to me like it avoids reading incorrect array elements almost by chance.

Typically it would also be necessary to check that the known object is really an array, but that should be guaranteed by `isArrayWithConstantElements()` :+1: